planner/cascades/transformation_rules.go

```diff
@@ -634,3 +637,66 @@ func (r *EliminateProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.
 	}
 	return finalGroupExprs, true, false, nil
 }
+
+// PushTopNDownProjection pushes TopN to Projection.
+type PushTopNDownProjection struct {
+	baseRule
+}
+
+// NewRulePushTopNDownProjection creates a new Transformation PushTopNDownProjection.
+// The pattern of this rule is `TopN->Projection->X` to `Projection->TopN->X`.
+func NewRulePushTopNDownProjection() Transformation {
+	rule := &PushTopNDownProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandTopN,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandProjection, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *PushTopNDownProjection) Match(expr *memo.ExprIter) bool {
+	proj := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	for _, expr := range proj.Exprs {
+		if expression.HasAssignSetVarFunc(expr) {
```

Comment:
Would you please add this check to the old planner as well?

Reply:
Sure. I will do it.

Reply:
```
func (p *LogicalProjection) pushDownTopN(topN *LogicalTopN) LogicalPlan {
	for _, expr := range p.Exprs {
		if expression.HasAssignSetVarFunc(expr) {
			return p
		}
	}
        ......
	return p
}
```
Is this correct?

Reply:
You can move this check to `Match()` directly.

Reply:
Yes, it's correct

Reply:
Do I need to add test cases for that? And how?

Reply:
You can add a test in `TestTopNPushDown` of `plan_suite_unexported_in.json`.